rancher2/resource_rancher2_machine_config_v2.go

```diff
@@ -38,29 +39,51 @@ func resourceRancher2MachineConfigV2Create(d *schema.ResourceData, meta interfac
 
 	d.SetId(newObj.ID)
 	d.Set("kind", newObj.TypeMeta.Kind)
-	stateConf := &resource.StateChangeConf{
-		Pending:    []string{},
-		Target:     []string{"active"},
-		Refresh:    machineConfigV2StateRefreshFunc(meta, newObj.ID, newObj.TypeMeta.Kind),
-		Timeout:    d.Timeout(schema.TimeoutCreate),
-		Delay:      1 * time.Second,
-		MinTimeout: 3 * time.Second,
-	}
-	_, waitErr := stateConf.WaitForState()
-	if waitErr != nil {
-		return fmt.Errorf("[ERROR] waiting for machine config (%s) to be active: %s", newObj.ID, waitErr)
+
+	err = waitForMachineConfigV2(d, meta.(*Config), d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return fmt.Errorf("[ERROR] waiting for machine config (%s) to be active: %s", newObj.ID, err)
 	}
 
 	return flattenMachineConfigV2(d, newObj)
 }
 
+func waitForMachineConfigV2(d *schema.ResourceData, config *Config, interval time.Duration) error {
+	log.Printf("[INFO] Waiting for state Machine Config V2 %s", d.Id())
+
+	ctx, cancel := context.WithTimeout(context.Background(), interval)
+	defer cancel()
+	for {
+		kind := d.Get("kind").(string)
+		_, err := getMachineConfigV2ByID(config, d.Id(), kind)
+		if err == nil {
+			return nil
+		}
+		log.Printf("[INFO] Retrying on error Refreshing Machine Config V2 %s: %v", d.Id(), err)
+		if IsNotFound(err) || IsForbidden(err) {
+			d.SetId("")
+			return fmt.Errorf("Machine Config V2 %s not found: %s", d.Id(), err)
+		}
+		if IsNotAccessibleByID(err) {
+			// Restarting clients to update RBAC
+			config.RestartClients()
```

Comment:
I do not understand why we need to restart the client to update RBAC. Is it because we cannot refresh the cache for RBAC or something in this case?  Is it the way Terraform works that I am not aware of? 
Excuse me if my questions do not make sense.

Reply:
In this case, the bug is a race condition. Terraform finishes creating the machine config object and starts trying to create the cluster _before_ Rancher has given GET method access to the machine config object for a standard user. RBAC is the service that checks those permissions and returns whether or not access is allowed (high level-y), so restarting RBAC here refreshes the service and enables RBAC to see that Rancher has given access to that obj. Then terraform will provision correctly.

Does that make sense? This is basically the root of my fix.